### PR TITLE
data: status: introduce --not-in-remote

### LINKED ISSUE #8761: cloud versioning: status command
`dvc status` or `dvc data status` should work with cloud-versioned remotes.

Edit: The scope for this ticket is limited to `version_aware` remotes.

### PR DESCRIPTION
**Pull request by efiop**
Slow POC implementation.

image

#8761

### CODE REVIEW

File: dvc/commands/data.py [hunk 1]
```diff
@@ -114,6 +117,8 @@ def run(self) -> int:
             status = self.repo.data_status(
                 granular=self.args.granular,
                 untracked_files=self.args.untracked_files,
+                not_in_cache=True,
+                not_in_remote=self.args.not_in_remote,
```
Comment (skshetry) [inline dvc/commands/data.py:121]: You might need to drop the `not_in_remote`/`not_in_cache` entries.

```python
if not self.args.not_in_remote:
	status.pop("not_in_remote")
```

**Comment by efiop** [comment]
For the record: reworked a few things to show `not in remote` by default, which is the most useful for users (both regular and cloud versioning remotes work and the code is the same, which is important). The hiccup is related to pre-requisites to finally using a persistent index here (for remotes specifically) so that we don't ring the cloud every time.

**Comment by efiop** [comment]
The current implementation uses persistent index (don't forget `dvc config feature.data_index_cache true` if you'll be trying it out) for cloud versioning, but not for regular remotes. Will add support there and get back here for `--remote-update` flag and some edge cases/cleanup.

**Comment by efiop** [comment]
This is taking way too long and there are still some questions on how to index http remotes (since listing is not part of the standard). Hate to do this, but I'll bring back `--not-in-remote` for now in the interest of time 🙁

**Comment by dberenbaum** [comment]
> This is taking way too long and there are still some questions on how to index http remotes (since listing is not part of the standard). Hate to do this, but I'll bring back `--not-in-remote` for now in the interest of time 🙁

I think it's fine to merge it this way. 

However, I did run into really slow performance when doing `dvc data status --granular --not-in-remote`. I haven't tested with persistent index (how do I enable it?). Here's what I did:

```bash
# Add 2800 files
$ dvc get git@github.com:iterative/dataset-registry.git use-cases/cats-dogs
$ dvc add cats-dogs
# Configure remote
$ dvc remote add -d -f versioned s3://dave-sandbox-versioning/test/versioned
$ dvc remote modify versioned version_aware true
# Get status
$ time dvc data status --granular --not-in-remote
...
dvc data status --granular --not-in-remote  12.77s user 1.42s system 14% cpu 1:37.57 total
```

I'm not even sure we actually need to bother with checking the cloud here. If there's no cloud version ID in the `.dvc` file, should we just report that it's not in remote?

**Comment by efiop** [comment]
@dberenbaum Great idea with filling from version_id's in dvcfiles!

Sorry for the pause here, trying to move persistent index to within the .dvc/ directory (currently it is global, which was done in the hopes to help with NFS use case but it turned out that wouldn't help), so that we could enable persistent index by default. The version_id trick would definitely work, but I'm a bit worried about cases like corrupted dvc files that would result in invalid percieved state. Even if we go with that trick, we'll still need `--remote-update` to be able to fix potential bad states and then we would again need to save to the persistent index to be able to use that later. But using that trick to do initial remote index fill-up sounds pretty good to me.

**Comment by dberenbaum** [comment]
> If there's no cloud version ID in the `.dvc` file, should we just report that it's not in remote?

Discussed with @efiop that for this specific scenario where the `version_id` is missing, an optimization would be to not check the remote at all. This should be safe to do since no matter what versions exist on the remote, we don't have a local `version_id` to compare. This is not a blocker if performance looks fine otherwise.

File: dvc/commands/data.py (hunk 1 above)
Comment (efiop) [reply dvc/commands/data.py:120]: You mean for json?

File: tests/func/test_data_status.py [hunk 2]
```diff
@@ -396,6 +398,42 @@ def test_missing_dir_object_from_index(M, tmp_dir, dvc, scm):
...
+    assert dvc.data_status(untracked_files="all", not_in_remote=True) == {
+        **EMPTY_STATUS,
+        "untracked": M.unordered("foobar.dvc", "dir.dvc", ".gitignore"),
+        "committed": {"added": M.unordered("foobar", join("dir", ""))},
+        "not_in_remote": M.unordered("foobar", join("dir", "")),
+        "git": M.dict(),
+    }
+
+    assert dvc.data_status(
+        granular=True, untracked_files="all", not_in_remote=True
```
Comment (skshetry) [inline tests/func/test_data_status.py:414]: `untracked_files="all"` may be unnecessary to test here.

**Pull request merged**